File: ios/Releaf/DesignSystem/Components/PaperSavedSheet.swift
/*
 * PaperSavedSheet.swift
 *
 * Bottom sheet that opens when the RE-LEAF eyebrow on the strip is
 * tapped. Shows the math behind the two numbers in the strip:
 *
 *   - per-kind table  (count × sheets/each = sheets)
 *   - total           (sum)
 *   - tree readout    (sheets ÷ sheetsPerTree)
 *   - honest closing  copy on what the number means
 *
 * Counts come from the page (or aggregate) the strip was built from.
 * The multipliers come from `AppMetrics.PaperPerCapture` — same source
 * that the strip itself uses, so the sub-totals always reconcile.
 *
 * The sheet uses `.presentationDetents([.medium, .large])` so the
 * caller doesn't have to pick a size; the system grows it on tall
 * notebooks where the table runs longer.
 */

import SwiftUI

public struct PaperSavedSheet: View {
    public let photos: Int
    public let voiceNotes: Int
    public let todoItems: Int
    public let scans: Int
    public let contacts: Int
    public let places: Int
    public let notes: Int
    /// Optional accent override for the eyebrow + total readout
    /// tones. Nil → defaults to the green theme. PageDetail
    /// passes the parent-notebook color so the explainer matches
    /// the surface that opened it. The breakdown-row glyphs are
    /// intentionally not overridden — they're per-capture-mode
    /// color codes (scan green, contact blue, etc.) and shouldn't
    /// morph with notebook color.
    public let accentOverride: Color?
    public let onClose: () -> Void

    public init(
        photos: Int = 0,
        voiceNotes: Int = 0,
        todoItems: Int = 0,
        scans: Int = 0,
        contacts: Int = 0,
        places: Int = 0,
        notes: Int = 0,
        accentOverride: Color? = nil,
        onClose: @escaping () -> Void
    ) {
        self.photos = photos
        self.voiceNotes = voiceNotes
        self.todoItems = todoItems
        self.scans = scans
        self.contacts = contacts
        self.places = places
        self.notes = notes
        self.accentOverride = accentOverride
        self.onClose = onClose
    }

    public var body: some View {
        let impact = ReleafImpact(
            photos: photos, voiceNotes: voiceNotes, todoItems: todoItems,
            scans: scans, contacts: contacts, places: places, notes: notes
        )
        let totalCaptures = photos + voiceNotes + todoItems + scans + contacts + places + notes

        let eyebrowTint = accentOverride ?? AppColors.themeGreenDeep
        ScrollView {
            VStack(alignment: .leading, spacing: AppSpacing.s3) {
                Text("RE-LEAF")
                    .font(AppText.eyebrow)
                    .tracking(AppLetterSpacing.eyebrow)
                    .foregroundStyle(eyebrowTint)

                Text("how paper saved is counted")
                    .font(AppText.editorialTitle)
                    .foregroundStyle(AppColors.textPrimary)
                    .padding(.bottom, AppSpacing.s2)

                summaryTiles(impact: impact, accentTint: eyebrowTint)
                    .padding(.bottom, AppSpacing.s3)

                Text("PER CAPTURE · ON THIS PAGE")
                    .font(AppText.eyebrow)
                    .tracking(AppLetterSpacing.eyebrow)
                    .foregroundStyle(AppColors.textSecondary)
                    .padding(.bottom, AppSpacing.s1)

                breakdownRows()

                totalRow(impact: impact, totalCaptures: totalCaptures)
                    .padding(.top, AppSpacing.s2)

                Text(rationale)
                    .font(AppText.meta)
                    .foregroundStyle(AppColors.textTertiary)
                    .lineSpacing(3)
                    .padding(.top, AppSpacing.s4)

                AppButton("Close", variant: .primary, action: onClose)
                    .padding(.top, AppSpacing.s3)
            }
            .padding(.horizontal, AppSpacing.s5)
            .padding(.top, AppSpacing.s4)
            .padding(.bottom, AppSpacing.s6)
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .background(AppColors.cardSolid.ignoresSafeArea())
        .presentationDetents([.medium, .large])
        .presentationDragIndicator(.visible)
    }

    // MARK: - Pieces

    private func summaryTiles(impact: ReleafImpact, accentTint: Color) -> some View {
        let readout = impact.treeReadout
        return HStack(spacing: AppSpacing.s2) {
            SummaryTile(
                eyebrow: "SHEETS",
                value: impact.formattedSheets,
                caption: "across this page",
                eyebrowTint: accentTint
            )
            SummaryTile(
                eyebrow: "TREES",
                value: readout.number,
                caption: readout.unit,
                eyebrowTint: AppColors.textSecondary
            )
        }
    }

    @ViewBuilder
    private func breakdownRows() -> some View {
        let m = AppMetrics.PaperPerCapture.self
        VStack(spacing: 0) {
            BreakdownRow(label: "scans",       count: scans,      multiplier: m.scan,
                         glyph: AppColors.green)
            BreakdownRow(label: "notes",       count: notes,      multiplier: m.note,
                         glyph: AppColors.themeGreenPrimary)
            BreakdownRow(label: "voice notes", count: voiceNotes, multiplier: m.voice,
                         glyph: AppColors.warning)
            BreakdownRow(label: "contacts",    count: contacts,   multiplier: m.contact,
                         glyph: AppColors.info)
            BreakdownRow(label: "places",      count: places,     multiplier: m.place,
                         glyph: AppColors.coralDeep)
            BreakdownRow(label: "photos",      count: photos,     multiplier: m.photo,
                         glyph: AppColors.themeGreenPrimary.opacity(0.7),
                         isLast: false)
            BreakdownRow(label: "to-do",       count: todoItems,  multiplier: m.todo,
                         glyph: AppColors.themeGreenPrimary.opacity(0.55),
                         isLast: true)
        }
    }

    private func totalRow(impact: ReleafImpact, totalCaptures: Int) -> some View {
        HStack {
            Text("total · \(totalCaptures) capture\(totalCaptures == 1 ? "" : "s")")
                .font(AppText.body)
                .foregroundStyle(AppColors.textPrimary)
            Spacer()
            Text(impact.formattedSheets)
                .font(.system(size: 22, design: .serif))
                .foregroundStyle(AppColors.green)
        }
        .padding(.vertical, AppSpacing.s3)
        .overlay(
            Rectangle()
                .fill(AppColors.borderStrong)
                .frame(height: 1),
            alignment: .top
        )
    }

    private var rationale: String {
        "a mature pine yields about \(Int(AppMetrics.sheetsPerTree).formatted(.number)) letter-size sheets. each capture is rated against what it would have replaced on paper — a scan against one printed page, a note against a quarter, a voice note against a tenth. small on its own. across years of notebooks, less so."
    }
}

// MARK: - Summary tile

private struct SummaryTile: View {
    let eyebrow: String
    let value: String
    let caption: String
    let eyebrowTint: Color

    var body: some View {
        VStack(alignment: .leading, spacing: AppSpacing.s1) {
            Text(eyebrow)
                .font(AppText.eyebrow)
                .tracking(AppLetterSpacing.eyebrow)
                .foregroundStyle(eyebrowTint)
            Text(value)
                .font(.system(size: 26, weight: .regular, design: .serif))
                .foregroundStyle(AppColors.textPrimary)
                .lineLimit(1)
                .minimumScaleFactor(0.6)
            Text(caption)
                .font(AppText.tag)
                .foregroundStyle(AppColors.textTertiary)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(AppSpacing.s3)
        .background(
            RoundedRectangle(cornerRadius: AppRadius.md, style: .continuous)
                .fill(AppColors.canvas)
        )
    }
}

// MARK: - Breakdown row

private struct BreakdownRow: View {
    let label: String
    let count: Int
    let multiplier: Double
    let glyph: Color
    var isLast: Bool = false

    var body: some View {
        let subtotal = Double(count) * multiplier
        VStack(spacing: 0) {
            HStack(alignment: .firstTextBaseline) {
                LeafDropletGlyph(tint: glyph, size: 11)
                    .padding(.trailing, AppSpacing.s1)
                Text(label)
                    .font(AppText.body)
                    .foregroundStyle(AppColors.textPrimary)
                Spacer()
                Text("\(count) × \(formatMultiplier(multiplier))")
                    .font(AppText.tag)
                    .foregroundStyle(AppColors.textTertiary)
                    .padding(.trailing, AppSpacing.s3)
                Text(String(format: "%.2f", subtotal))
                    .font(.system(size: 15, design: .serif))
                    .foregroundStyle(AppColors.textPrimary)
                    .frame(minWidth: 44, alignment: .trailing)
            }
            .padding(.vertical, AppSpacing.s2)
            if !isLast {
                Rectangle()
                    .fill(AppColors.borderDefault)
                    .frame(height: 0.5)
            }
        }
    }

    private func formatMultiplier(_ value: Double) -> String {
        // Trim trailing zeros: 1.00 → 1.0, 0.30 → 0.3.
        let formatted = String(format: "%.2f", value)
        if formatted.hasSuffix("0") {
            return String(formatted.dropLast())
        }
        return formatted
    }
}

#if DEBUG
struct PaperSavedSheet_Previews: PreviewProvider {
    static var previews: some View {
        Color.gray.opacity(0.2)
            .sheet(isPresented: .constant(true)) {
                PaperSavedSheet(
                    photos: 2, voiceNotes: 1, todoItems: 1,
                    scans: 2, contacts: 1, places: 1, notes: 1,
                    onClose: {}
                )
            }
    }
}
#endif

File: apps/releaf/android/app/src/main/java/app/releaf/mobile/ui/components/PaperSavedSheet.kt
/*
 * PaperSavedSheet.kt
 *
 * ModalBottomSheet that opens when the RE-LEAF eyebrow on the strip is
 * tapped. Shows the math behind the two numbers in the strip:
 *
 *   - per-kind table  (count × sheets/each = sheets)
 *   - total           (sum)
 *   - tree readout    (sheets ÷ sheetsPerTree)
 *   - honest closing  copy on what the number means
 *
 * Mirrors `PaperSavedSheet.swift` (iOS).
 */

package app.releaf.mobile.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import app.releaf.mobile.ui.theme.AppColors
import app.releaf.mobile.ui.theme.AppMetrics
import app.releaf.mobile.ui.theme.AppRadius
import app.releaf.mobile.ui.theme.AppSpacing
import app.releaf.mobile.ui.theme.AppTypography
import app.releaf.mobile.ui.theme.ReleafImpact
import java.text.NumberFormat
import java.util.Locale

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PaperSavedSheet(
    photos: Int,
    voiceNotes: Int,
    todoItems: Int,
    scans: Int,
    contacts: Int,
    places: Int,
    notes: Int,
    onDismiss: () -> Unit,
    /** Optional accent override for the eyebrow + summary tile
     *  tones. Nil → defaults to the green theme. PageDetail
     *  passes the parent-notebook color so the explainer matches
     *  the surface that opened it. Per-capture-mode glyphs in
     *  the breakdown rows stay fixed (semantic color codes). */
    accentOverride: androidx.compose.ui.graphics.Color? = null,
) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false)

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState       = sheetState,
        containerColor   = AppColors.CardSolid,
        contentColor     = AppColors.TextPrimary,
        shape            = RoundedCornerShape(topStart = 24.dp, topEnd = 24.dp),
    ) {
        SheetBody(
            photos = photos, voiceNotes = voiceNotes, todoItems = todoItems,
            scans = scans, contacts = contacts, places = places, notes = notes,
            accentOverride = accentOverride,
            onClose = onDismiss,
        )
    }
}

@Composable
private fun SheetBody(
    photos: Int,
    voiceNotes: Int,
    todoItems: Int,
    scans: Int,
    contacts: Int,
    places: Int,
    notes: Int,
    accentOverride: androidx.compose.ui.graphics.Color?,
    onClose: () -> Unit,
) {
    val impact = ReleafImpact.from(
        photos = photos, voiceNotes = voiceNotes, todoItems = todoItems,
        scans = scans, contacts = contacts, places = places, notes = notes,
    )
    val totalCaptures = photos + voiceNotes + todoItems + scans + contacts + places + notes
    val scroll = rememberScrollState()

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .verticalScroll(scroll)
            .padding(horizontal = AppSpacing.s5)
            .padding(top = AppSpacing.s4, bottom = AppSpacing.s6),
        verticalArrangement = Arrangement.spacedBy(AppSpacing.s3),
    ) {
        val eyebrowTint = accentOverride ?: AppColors.ThemeGreenDeep
        Text("RE-LEAF",                style = AppTypography.Eyebrow,        color = eyebrowTint)
        Text("how paper saved is counted", style = AppTypography.EditorialTitle, color = AppColors.TextPrimary)

        Spacer(Modifier.height(AppSpacing.s1))

        SummaryTiles(impact = impact, accentTint = eyebrowTint)

        Spacer(Modifier.height(AppSpacing.s2))

        Text(
            "PER CAPTURE · ON THIS PAGE",
            style = AppTypography.Eyebrow,
            color = AppColors.TextSecondary,
        )

        BreakdownRows(
            scans      = scans,
            notes      = notes,
            voiceNotes = voiceNotes,
            contacts   = contacts,
            places     = places,
            photos     = photos,
            todoItems  = todoItems,
        )

        TotalRow(impact = impact, totalCaptures = totalCaptures)

        Spacer(Modifier.height(AppSpacing.s2))

        Text(
            text  = rationale(),
            style = AppTypography.Meta,
            color = AppColors.TextTertiary,
        )

        AppButton(
            "Close",
            onClick   = onClose,
            variant   = AppButtonVariant.Primary,
            modifier  = Modifier.padding(top = AppSpacing.s3),
        )
    }
}

@Composable
private fun SummaryTiles(impact: ReleafImpact, accentTint: Color) {
    val readout = impact.treeReadout
    Row(horizontalArrangement = Arrangement.spacedBy(AppSpacing.s2)) {
        SummaryTile(
            eyebrow     = "SHEETS",
            value       = impact.formattedSheets,
            caption     = "across this page",
            eyebrowTint = accentTint,
            modifier    = Modifier.weight(1f),
        )
        SummaryTile(
            eyebrow     = "TREES",
            value       = readout.number,
            caption     = readout.unit,
            eyebrowTint = AppColors.TextSecondary,
            modifier    = Modifier.weight(1f),
        )
    }
}

@Composable
private fun SummaryTile(
    eyebrow: String,
    value: String,
    caption: String,
    eyebrowTint: Color,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier
            .clip(RoundedCornerShape(AppRadius.md))
            .background(AppColors.Canvas)
            .padding(AppSpacing.s3),
        verticalArrangement = Arrangement.spacedBy(AppSpacing.s1),
    ) {
        Text(eyebrow, style = AppTypography.Eyebrow,        color = eyebrowTint)
        Text(value,   style = AppTypography.EditorialTitle, color = AppColors.TextPrimary, maxLines = 1)
        Text(caption, style = AppTypography.Tag,            color = AppColors.TextTertiary)
    }
}

@Composable
private fun BreakdownRows(
    scans: Int,
    notes: Int,
    voiceNotes: Int,
    contacts: Int,
    places: Int,
    photos: Int,
    todoItems: Int,
) {
    val m = AppMetrics.PaperPerCapture
    Column {
        BreakdownRow("scans",       scans,      m.Scan,    AppColors.Green)
        BreakdownRow("notes",       notes,      m.Note,    AppColors.ThemeGreenPrimary)
        BreakdownRow("voice notes", voiceNotes, m.Voice,   AppColors.Warning)
        BreakdownRow("contacts",    contacts,   m.Contact, AppColors.Info)
        BreakdownRow("places",      places,     m.Place,   AppColors.CoralDeep)
        BreakdownRow("photos",      photos,     m.Photo,   AppColors.ThemeGreenPrimary.copy(alpha = 0.7f))
        BreakdownRow("to-do",       todoItems,  m.Todo,    AppColors.ThemeGreenPrimary.copy(alpha = 0.55f), isLast = true)
    }
}

@Composable
private fun BreakdownRow(
    label: String,
    count: Int,
    multiplier: Double,
    glyph: Color,
    isLast: Boolean = false,
) {
    val subtotal = count * multiplier
    Column {
        Row(
            modifier              = Modifier.padding(vertical = AppSpacing.s2),
            verticalAlignment     = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(AppSpacing.s1),
        ) {
            LeafDropletGlyph(tint = glyph)
            Text(label, style = AppTypography.Body, color = AppColors.TextPrimary, modifier = Modifier.weight(1f))
            Text(
                text     = "$count × ${formatMultiplier(multiplier)}",
                style    = AppTypography.Tag,
                color    = AppColors.TextTertiary,
                modifier = Modifier.padding(end = AppSpacing.s3),
            )
            Text(
                text     = String.format(Locale.US, "%.2f", subtotal),
                style    = AppTypography.Body,
                color    = AppColors.TextPrimary,
                textAlign = TextAlign.End,
                modifier = Modifier.widthIn(min = 44.dp),
            )
        }
        if (!isLast) {
            Box(modifier = Modifier.fillMaxWidth().height(0.5.dp).background(AppColors.BorderDefault))
        }
    }
}

@Composable
private fun TotalRow(impact: ReleafImpact, totalCaptures: Int) {
    Column {
        Box(modifier = Modifier.fillMaxWidth().height(1.dp).background(AppColors.BorderStrong))
        Row(
            modifier              = Modifier.padding(vertical = AppSpacing.s3).fillMaxWidth(),
            verticalAlignment     = Alignment.CenterVertically,
        ) {
            Text(
                text     = "total · $totalCaptures capture${if (totalCaptures == 1) "" else "s"}",
                style    = AppTypography.Body,
                color    = AppColors.TextPrimary,
                modifier = Modifier.weight(1f),
            )
            Text(
                text  = impact.formattedSheets,
                style = AppTypography.EditorialTitle,
                color = AppColors.Green,
            )
        }
    }
}

private fun rationale(): String {
    val sheets = NumberFormat.getInstance(Locale.US).format(AppMetrics.sheetsPerTree.toLong())
    return "a mature pine yields about $sheets letter-size sheets. " +
        "each capture is rated against what it would have replaced on " +
        "paper — a scan against one printed page, a note against a " +
        "quarter, a voice note against a tenth. small on its own. " +
        "across years of notebooks, less so."
}

private fun formatMultiplier(value: Double): String {
    val formatted = String.format(Locale.US, "%.2f", value)
    return if (formatted.endsWith("0")) formatted.dropLast(1) else formatted
}
